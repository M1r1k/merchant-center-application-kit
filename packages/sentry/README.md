# @commercetools-frontend/sentry

Components to set up Sentry for a MC application.

## Install

```bash
$ npm install --save @commercetools-frontend/sentry
```
