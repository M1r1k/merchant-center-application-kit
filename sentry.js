import Raven from 'raven-js';

export const boot = () => {
  if (window.app.env === 'production')
    Raven.config(window.app.tracking.sentry, {
      release: window.app.revision,
      tags: { role: 'frontend' },
      environment: window.app.env,
      // in order to reduce the noise in sentry we only track errors that come
      // from our code and ignore errors that come from e.g. intercom
      // https://blog.sentry.io/2017/03/27/tips-for-reducing-javascript-error-noise.html
      whitelistUrls: [window.app.cdnUrl, window.app.frontendHost],
    }).install();
};

export const updateUser = user => {
  if (user && window.app.env === 'production') {
    // to avoid sending personal data to sentry we anonymize the email address
    // by only sending the domain part or the email
    const emailTld = user.email.split('@')[1];
    Raven.setUserContext({
      email: `xxx@${emailTld}`,
      id: user.id,
    });
  }
};

export const stopTrackingUser = () => {
  if (window.app.env === 'production') Raven.setUserContext();
};

// NOTE: only exported for tests => use reportError below for your code
export const createSentryErrorReporter = (
  environment,
  debugLogger = console,
  productionLogger = Raven
) => (error, extraInfo) => {
  if (error instanceof Error === false && environment === 'development') {
    debugLogger.warn(
      '[SENTRY]: You called "sentry.reportError" with an argument that is not an instance of "Error". ' +
        '"Error" should be preferred so that a stack-trace can be made available in sentry. ' +
        'See: https://docs.sentry.io/clients/javascript/usage/#try-catch'
    );
  }

  if (environment === 'production') {
    // logs error to sentry
    if (error instanceof Error) {
      productionLogger.captureException(error, extraInfo);
    } else {
      productionLogger.captureMessage(error, extraInfo);
    }

    // Generate a unique ID referring to the last generated Sentry error
    const errorId = productionLogger.lastEventId();

    // The error stack should be available in Sentry, so there is no
    // need to print it in the console as well.
    // We just notify that an error occurred and provide the error ID.
    debugLogger.error(`[SENTRY]: An error occured (ID: ${errorId}).`);
  } else {
    debugLogger.error('[SENTRY]:', error);
  }
};

export const reportErrorToSentry = createSentryErrorReporter(
  window.app.env || process.env.NODE_ENV
);
