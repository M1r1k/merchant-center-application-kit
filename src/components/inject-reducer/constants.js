// eslint-disable-next-line import/prefer-default-export
export const ACTIVATE_PLUGIN = 'ACTIVATE_PLUGIN';
