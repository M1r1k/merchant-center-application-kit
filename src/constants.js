// eslint-disable-next-line import/prefer-default-export
export const USER_PROFILE_FORM_NAME =
  'packages-shared/app-shell/USER_PROFILE_FORM_NAME';
