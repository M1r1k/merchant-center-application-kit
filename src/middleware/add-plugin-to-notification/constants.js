export const __LOCAL = 'LOCAL';
export const __GLOBAL = 'GLOBAL';
