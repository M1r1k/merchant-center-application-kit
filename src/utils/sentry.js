import Raven from 'raven-js';

export const boot = () => {
  if (window.app.env === 'production')
    Raven.config(window.app.tracking.sentry, {
      release: window.app.revision,
      tags: { role: 'frontend' },
      environment: window.app.env,
    }).install();
};

export const updateUser = user => {
  if (user && window.app.env === 'production')
    Raven.setUserContext({
      email: user.email,
      id: user.id,
    });
};

export const stopTrackingUser = () => {
  if (window.app.env === 'production') Raven.setUserContext();
};

// NOTE: only exported for tests => use reportError below for your code
export const createErrorReporter = (
  environment,
  debugLogger = console,
  productionLogger = Raven
) => error => {
  if (error instanceof Error === false && environment === 'development') {
    debugLogger.warn(
      '[SENTRY]: You called "sentry.reportError" with an argument that is not an instance of "Error". ' +
        '"Error" should be preferred so that a stack-trace can be made available in sentry. ' +
        'See: https://docs.sentry.io/clients/javascript/usage/#try-catch'
    );
  }

  if (environment === 'production') {
    // logs error to sentry
    if (error instanceof Error) {
      productionLogger.captureException(error);
    } else {
      productionLogger.captureMessage(error);
    }

    // Generate a unique ID referring to the last generated Sentry error
    const errorId = productionLogger.lastEventId();

    // The error stack should be available in Sentry, so there is no
    // need to print it in the console as well.
    // We just notify that an error occurred and provide the error ID.
    debugLogger.error(`[SENTRY]: An error occured (ID: ${errorId}).`);
  } else if (environment === 'development') {
    debugLogger.error('[SENTRY]:', error);
  }
};

export const reportError = createErrorReporter(
  window.app.env || process.env.NODE_ENV
);
