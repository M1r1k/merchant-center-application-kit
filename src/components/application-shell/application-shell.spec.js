import React from 'react';
import { shallow } from 'enzyme';
import { ReconfigureFlopFlip } from '@flopflip/react-broadcast';
import { DOMAINS } from '@commercetools-local/constants';
import { reportErrorToSentry } from '@commercetools-local/sentry';
import ConfigureIntlProvider from '../configure-intl-provider';
import ProjectContainer from '../project-container';
import FetchUser from '../fetch-user';
import IntercomBooter from '../intercom-booter';
import IntercomUserTracker from '../intercom-user-tracker';
import NavBar from '../navbar';
import AsyncUserProfile from '../user-profile/async';
import ApplicationShell, {
  RestrictedApplication,
  UnrestrictedApplication,
} from './application-shell';

jest.mock('@commercetools-local/utils/storage');
jest.mock('@commercetools-local/sentry');

const createTestProps = props => ({
  i18n: {
    en: { title: 'Title en' },
    'en-US': { title: 'Title' },
    de: { title: 'Titel' },
  },
  configuration: {},
  trackingEventWhitelist: {},
  render: jest.fn(),
  notificationsByDomain: {
    global: [],
    page: [],
    side: [],
  },
  showNotification: jest.fn(),
  mapPluginNotificationToComponent: jest.fn(),
  showApiErrorNotification: jest.fn(),
  showUnexpectedErrorNotification: jest.fn(),
  onRegisterErrorListeners: jest.fn(),
  INTERNAL__isApplicationFallback: false,
  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<ApplicationShell {...props} />);
  });
  describe('providers', () => {
    it('should pass "configuration" to <ConfigurationProvider>', () => {
      expect(wrapper.find('ConfigurationProvider')).toHaveProp(
        'configuration',
        props.configuration
      );
    });
  });
  describe('trackers', () => {
    it('should render <IntercomUrlTracker> below <Router>', () => {
      expect(wrapper).toRender('Router > IntercomUrlTracker');
    });
    it('should render <GtmBooter> below <IntercomUrlTracker>', () => {
      expect(wrapper).toRender('IntercomUrlTracker > GtmBooter');
    });
    it('should pass "trackingEventWhitelist" to <GtmBooter>', () => {
      expect(wrapper.find('GtmBooter')).toHaveProp(
        'trackingEventWhitelist',
        props.trackingEventWhitelist
      );
    });
    it('should render <Switch> after track components', () => {
      expect(wrapper).toRender('GtmBooter > Switch');
    });
  });
  it('should render <Route> for "/logout"', () => {
    expect(wrapper).toRender({ path: '/logout' });
  });

  describe('<Authenticated>', () => {
    let routeRenderWrapper;
    let authRenderWrapper;
    beforeEach(() => {
      routeRenderWrapper = shallow(
        <div>
          {wrapper
            .find('Switch > Route')
            .last()
            .prop('render')()}
        </div>
      );
    });
    describe('when user is authenticated', () => {
      beforeEach(() => {
        authRenderWrapper = shallow(
          <div>
            {routeRenderWrapper.find('Authenticated').prop('children')({
              isAuthenticated: true,
            })}
          </div>
        );
      });
      it('should render <RestrictedApplication> after track components', () => {
        expect(authRenderWrapper).toRender('RestrictedApplication');
      });
    });

    describe('when user is not authenticated', () => {
      beforeEach(() => {
        authRenderWrapper = shallow(
          <div>
            {routeRenderWrapper.find('Authenticated').prop('children')({
              isAuthenticated: false,
            })}
          </div>
        );
      });
      it('should pass "locale" to <ConfigureIntlProvider>', () => {
        expect(authRenderWrapper.find(ConfigureIntlProvider)).toHaveProp(
          'locale',
          'en-US'
        );
      });
      it('should pass "messages" to <ConfigureIntlProvider>', () => {
        expect(authRenderWrapper.find(ConfigureIntlProvider)).toHaveProp(
          'messages',
          { title: 'Title en' }
        );
      });
      it('should render <UnrestrictedApplication> after track components', () => {
        expect(authRenderWrapper).toRender('UnrestrictedApplication');
      });
    });
  });
});
describe('<RestrictedApplication>', () => {
  let props;
  let wrapper;
  let userData;
  let fetchUserWrapper;
  describe('rendering', () => {
    beforeEach(() => {
      props = createTestProps();
      const rootWrapper = shallow(<RestrictedApplication {...props} />);
      userData = {
        isLoading: false,
        user: {
          id: 'u1',
          email: 'john.snow@got.com',
          firstName: 'John',
          lastName: 'Snow',
          availableProjects: [],
          language: 'en-US',
        },
      };
      wrapper = shallow(
        <div>{rootWrapper.find(FetchUser).prop('children')(userData)}</div>
      );
    });
    describe('when fetching the user returns an error', () => {
      beforeEach(() => {
        reportErrorToSentry.mockClear();
        props = createTestProps();
        const rootWrapper = shallow(<RestrictedApplication {...props} />);
        userData = {
          isLoading: false,
          error: new Error('Failed to fetch'),
        };
        fetchUserWrapper = shallow(
          <div>{rootWrapper.find(FetchUser).prop('children')(userData)}</div>
        );
      });
      it('should pass "locale" to <ConfigureIntlProvider>', () => {
        expect(fetchUserWrapper.find(ConfigureIntlProvider)).toHaveProp(
          'locale',
          'en-US'
        );
      });
      it('should render <ErrorApologizer>', () => {
        expect(fetchUserWrapper).toRender('ErrorApologizer');
      });
      it('should report error to sentry', () => {
        expect(reportErrorToSentry).toHaveBeenCalledWith(
          new Error('Failed to fetch'),
          {}
        );
      });
    });
    it('should match layout structure', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should boot intercom', () => {
      expect(wrapper).toRender(IntercomBooter);
    });
    it('should pass user "locale" to <ConfigureIntlProvider>', () => {
      expect(fetchUserWrapper.find(ConfigureIntlProvider)).toHaveProp(
        'locale',
        userData.user.language
      );
    });
    it('should pass "messages" to <ConfigureIntlProvider>', () => {
      expect(fetchUserWrapper.find(ConfigureIntlProvider)).toHaveProp(
        'messages',
        { title: 'Title en' }
      );
    });
    describe('layout', () => {
      it('should render "global-notifications" container inside "app-layout"', () => {
        expect(wrapper).toRender('.app-layout > .global-notifications');
      });
      it('should render "header" element inside "app-layout"', () => {
        expect(wrapper).toRender('.app-layout > header');
      });
      it('should render "aside" element inside "app-layout"', () => {
        expect(wrapper).toRender('.app-layout > aside');
      });
      it('should render "main" container inside "app-layout"', () => {
        expect(wrapper).toRender('.app-layout > .main');
      });
      it('should mark "main" container with "main" role', () => {
        expect(wrapper).toRender('.app-layout > .main[role="main"]');
      });
    });
    it('should render GLOBAL <NotificationsList>', () => {
      expect(wrapper.find('NotificationsList').at(0)).toHaveProp(
        'domain',
        DOMAINS.GLOBAL
      );
    });
    it('should render PAGE <NotificationsList>', () => {
      expect(wrapper.find('NotificationsList').at(1)).toHaveProp(
        'domain',
        DOMAINS.PAGE
      );
    });
    it('should render SIDE <NotificationsList>', () => {
      expect(wrapper.find('NotificationsList').at(2)).toHaveProp(
        'domain',
        DOMAINS.SIDE
      );
    });
    it('should render <AppBar> below header element', () => {
      expect(wrapper).toRender('header > AppBar');
    });
    it('should render <WithProjectKey> below aside element', () => {
      expect(wrapper.find('aside > WithProjectKey')).toHaveProp(
        'user',
        userData.user
      );
    });
    describe('<NavBar>', () => {
      let routeRenderWrapper;
      beforeEach(() => {
        routeRenderWrapper = shallow(
          <div>
            {wrapper.find('aside > WithProjectKey').prop('render')({
              routerProps: { location: {} },
              projectKey: 'foo-1',
            })}
          </div>
        );
      });
      it('should render <NavBar> inside <WithProjectKey> below aside element', () => {
        expect(routeRenderWrapper).toRender(NavBar);
      });
      it('should pass the projectKey matched from the URL', () => {
        expect(routeRenderWrapper.find(NavBar)).toHaveProp(
          'projectKey',
          'foo-1'
        );
      });
      it('should pass "useFullRedirectsForLinks"', () => {
        expect(routeRenderWrapper.find(NavBar)).toHaveProp(
          'useFullRedirectsForLinks',
          props.INTERNAL__isApplicationFallback
        );
      });
    });
    it('should render <Route> for "/account" below main container', () => {
      expect(wrapper.find('.main')).toRender({ path: '/account' });
    });
    describe('account routes', () => {
      let routeRenderWrapper;
      beforeEach(() => {
        routeRenderWrapper = shallow(
          <div>
            {wrapper
              .find('.main')
              .find({ path: '/account' })
              .prop('render')({ match: { url: '/account', path: '/account' } })}
          </div>
        );
      });
      it('should render <Route> for "/account/profile"', () => {
        expect(routeRenderWrapper).toRender({ path: '/account/profile' });
      });
      it('should render <Route> for redurect to "/account/profile"', () => {
        expect(routeRenderWrapper).toRender({ to: '/account/profile' });
      });
      describe('<AsyncUserProfile>', () => {
        let userProfileRouteRenderWrapper;
        beforeEach(() => {
          userProfileRouteRenderWrapper = shallow(
            <div>
              {routeRenderWrapper
                .find({ path: '/account/profile' })
                .prop('render')()}
            </div>
          );
        });
        it('should render <AsyncUserProfile> inside <Route> below main container', () => {
          expect(userProfileRouteRenderWrapper).toRender(AsyncUserProfile);
        });
      });
    });
    it('should render <Route> matching exact ":projectKey" path', () => {
      expect(wrapper.find('.main')).toRender({
        exact: true,
        path: '/:projectKey',
      });
    });
    describe('<Redirect> to dashboard', () => {
      let routeRenderWrapper;
      beforeEach(() => {
        routeRenderWrapper = shallow(
          <div>
            {wrapper
              .find('.main')
              .find({ exact: true, path: '/:projectKey' })
              .prop('render')({ match: { url: '/foo-1' } })}
          </div>
        );
      });
      it('should render <PageRedirect> to "/dashboard"', () => {
        expect(routeRenderWrapper.find('PageRedirect')).toHaveProp(
          'to',
          '/foo-1/dashboard'
        );
      });
      it('should render <PageRedirect> with "reload" to true', () => {
        expect(routeRenderWrapper.find('PageRedirect')).toHaveProp(
          'reload',
          true
        );
      });
    });
    it('should render <Route> matching ":projectKey" path', () => {
      expect(wrapper.find('.main')).toRender({
        exact: false,
        path: '/:projectKey',
      });
    });
    describe('project container <Route>', () => {
      let routeRenderWrapper;
      let routerProps;
      beforeEach(() => {
        routerProps = {
          location: { pathname: '/test-project/products' },
          match: { params: { projectKey: 'foo-1' } },
        };
        routeRenderWrapper = shallow(
          <div>
            {wrapper
              .find('.main')
              .find({ exact: false, path: '/:projectKey' })
              .prop('render')(routerProps)}
          </div>
        );
      });
      it('should match layout structure', () => {
        expect(wrapper).toMatchSnapshot();
      });
      it('should pass "projectKey" to <IntercomUserTracker>', () => {
        expect(routeRenderWrapper.find(IntercomUserTracker)).toHaveProp(
          'projectKey',
          routerProps.match.params.projectKey
        );
      });
      it('should pass "match" to <ProjectContainer>', () => {
        expect(routeRenderWrapper.find(ProjectContainer)).toHaveProp(
          'match',
          routerProps.match
        );
      });
      it('should pass "render" to <ProjectContainer>', () => {
        expect(routeRenderWrapper.find(ProjectContainer)).toHaveProp(
          'render',
          props.render
        );
      });
      it('should render <ReconfigureFlopflip>', () => {
        expect(routeRenderWrapper).toRender(ReconfigureFlopFlip);
      });

      it('should pass `projectKey` within `user` to `<ReconfigureFlopflip>`', () => {
        expect(routeRenderWrapper.find(ReconfigureFlopFlip)).toHaveProp(
          'user',
          expect.objectContaining({
            custom: expect.objectContaining({
              project: routerProps.match.params.projectKey,
            }),
          })
        );
      });
    });
    it('should render <Route> matching "/" path', () => {
      expect(wrapper.find('.main')).toRender({
        path: '/',
      });
    });
  });
});

describe('<UnrestrictedApplication>', () => {
  let props;
  let wrapper;
  describe('rendering', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<UnrestrictedApplication {...props} />);
    });
    it('should match layout structure', () => {
      expect(wrapper).toMatchSnapshot();
    });
    describe('catch <Route>', () => {
      let routeRenderWrapper;
      let routerProps;
      beforeEach(() => {
        routerProps = {
          location: { pathname: '/' },
        };
        routeRenderWrapper = shallow(
          <div>
            {wrapper
              .find('Route')
              .last()
              .prop('render')(routerProps)}
          </div>
        );
      });
      it('should render <Redirect> to "/login', () => {
        expect(routeRenderWrapper.find('Redirect')).toHaveProp(
          'to',
          expect.objectContaining({ pathname: '/login' })
        );
      });
      it('should render <Redirect> with "reason" in search param', () => {
        expect(routeRenderWrapper.find('Redirect')).toHaveProp(
          'to',
          expect.objectContaining({
            query: expect.objectContaining({ reason: 'unauthorized' }),
          })
        );
      });
      describe('when location pathname is "/"', () => {
        it('should render <Redirect> without "redirectTo" search param', () => {
          expect(routeRenderWrapper.find('Redirect')).not.toHaveProp(
            'to',
            expect.objectContaining({
              query: expect.objectContaining({
                redirectTo: expect.any(String),
              }),
            })
          );
        });
      });
      describe('when location pathname is "/foo-1/products"', () => {
        beforeEach(() => {
          routerProps = {
            location: { pathname: '/foo-1/products' },
          };
          routeRenderWrapper = shallow(
            <div>
              {wrapper
                .find('Route')
                .last()
                .prop('render')(routerProps)}
            </div>
          );
        });
        it('should render <Redirect> wit "redirectTo" search param', () => {
          expect(routeRenderWrapper.find('Redirect')).toHaveProp(
            'to',
            expect.objectContaining({
              query: expect.objectContaining({
                redirectTo: '/foo-1/products',
              }),
            })
          );
        });
      });
    });
  });
});

describe('lifecycle', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<ApplicationShell {...props} />);
  });
  describe('componentDidMount', () => {
    beforeEach(() => {
      wrapper.instance().componentDidMount();
    });
    it('should call onRegisterErrorListeners', () => {
      expect(props.onRegisterErrorListeners).toHaveBeenCalled();
    });
  });
});
