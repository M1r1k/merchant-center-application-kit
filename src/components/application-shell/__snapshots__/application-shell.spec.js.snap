// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RestrictedApplication> rendering project container <Route> should match layout structure 1`] = `
<div>
  <React.Fragment>
    <IntercomBooter
      showNotification={[MockFunction]}
    />
    <SentryUserTracker
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <GtmUserTracker
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <branch(IntercomUserTracker)
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <div
      className="app-layout"
    >
      <div
        className="global-notifications"
      >
        <NotificationsList
          domain="global"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
          showUnexpectedErrorNotification={[MockFunction]}
        />
      </div>
      <header>
        <AppBar
          user={
            Object {
              "availableProjects": Array [],
              "email": "john.snow@got.com",
              "firstName": "John",
              "id": "u1",
              "lastName": "Snow",
            }
          }
        />
      </header>
      <aside>
        <WithProjectKey
          render={[Function]}
          user={
            Object {
              "availableProjects": Array [],
              "email": "john.snow@got.com",
              "firstName": "John",
              "id": "u1",
              "lastName": "Snow",
            }
          }
        />
      </aside>
      <div
        className="main"
        role="main"
      >
        <NotificationsList
          domain="page"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
        />
        <NotificationsList
          domain="side"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
        />
        <Switch>
          <Route
            component={[Function]}
            path="/logout"
          />
          <Redirect
            from="/profile"
            push={false}
            to="/account/profile"
          />
          <Route
            path="/account"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/:projectKey"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          <Route
            exact={false}
            path="/:projectKey"
            render={[Function]}
          />
        </Switch>
      </div>
    </div>
  </React.Fragment>
</div>
`;

exports[`<RestrictedApplication> rendering should match layout structure 1`] = `
<div>
  <React.Fragment>
    <IntercomBooter
      showNotification={[MockFunction]}
    />
    <SentryUserTracker
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <GtmUserTracker
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <branch(IntercomUserTracker)
      user={
        Object {
          "availableProjects": Array [],
          "email": "john.snow@got.com",
          "firstName": "John",
          "id": "u1",
          "lastName": "Snow",
        }
      }
    />
    <div
      className="app-layout"
    >
      <div
        className="global-notifications"
      >
        <NotificationsList
          domain="global"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
          showUnexpectedErrorNotification={[MockFunction]}
        />
      </div>
      <header>
        <AppBar
          user={
            Object {
              "availableProjects": Array [],
              "email": "john.snow@got.com",
              "firstName": "John",
              "id": "u1",
              "lastName": "Snow",
            }
          }
        />
      </header>
      <aside>
        <WithProjectKey
          render={[Function]}
          user={
            Object {
              "availableProjects": Array [],
              "email": "john.snow@got.com",
              "firstName": "John",
              "id": "u1",
              "lastName": "Snow",
            }
          }
        />
      </aside>
      <div
        className="main"
        role="main"
      >
        <NotificationsList
          domain="page"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
        />
        <NotificationsList
          domain="side"
          mapPluginNotificationToComponent={[MockFunction]}
          notifications={Array []}
        />
        <Switch>
          <Route
            component={[Function]}
            path="/logout"
          />
          <Redirect
            from="/profile"
            push={false}
            to="/account/profile"
          />
          <Route
            path="/account"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/:projectKey"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          <Route
            exact={false}
            path="/:projectKey"
            render={[Function]}
          />
        </Switch>
      </div>
    </div>
  </React.Fragment>
</div>
`;

exports[`<UnrestrictedApplication> rendering should match layout structure 1`] = `
<Switch>
  <Route
    component={[Function]}
    path="/login/sso/callback"
  />
  <Route
    component={[Function]}
    path="/login/sso"
  />
  <Route
    component={[Function]}
    path="/login/locked"
  />
  <Route
    component={[Function]}
    path="/login"
  />
  <Route
    render={[Function]}
  />
</Switch>
`;

exports[`rendering should match main structure 1`] = `
<ConfigurationProvider
  configuration={Object {}}
>
  <ApolloProvider
    client={
      ApolloClient {
        "cache": InMemoryCache {
          "addTypename": true,
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": IntrospectionFragmentMatcher {
              "isReady": true,
              "match": [Function],
              "possibleTypesMap": Object {},
            },
            "storeFactory": [Function],
          },
          "data": ObjectCache {
            "data": Object {},
          },
          "optimistic": Array [],
          "silenceBroadcast": false,
          "watches": Array [],
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "resetStoreCallbacks": Array [],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": IntrospectionFragmentMatcher {
                "isReady": true,
                "match": [Function],
                "possibleTypesMap": Object {},
              },
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "watches": Array [],
          },
        },
        "version": "2.2.1",
        "watchQuery": [Function],
      }
    }
  >
    <branch(ConfigureIntlProvider)
      i18n={
        Object {
          "de": Object {},
          "en": Object {},
        }
      }
    >
      <React.Fragment>
        <VersionCheckSubscriber />
        <Connect(RequestsInFlightLoader) />
        <BrowserRouter>
          <IntercomUrlTracker>
            <GtmBooter
              trackingEventWhitelist={Object {}}
            >
              <Authenticated />
            </GtmBooter>
          </IntercomUrlTracker>
        </BrowserRouter>
      </React.Fragment>
    </branch(ConfigureIntlProvider)>
  </ApolloProvider>
</ConfigurationProvider>
`;
